src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if stream_id >= self.pcm_infos.as_ref().unwrap().len() as u32 {
+            return Err(Error::InvalidParam);
+        }
+        let pcm_info = &self.pcm_infos.as_ref().unwrap()[stream_id as usize];
+        Ok(pcm_info.channels_min..=pcm_info.channels_max)
+    }
+
+    /// Get features that a stream supports.
+    pub fn features_supported(&mut self, stream_id: u32) -> Result<PcmFeatures> {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if stream_id >= self.pcm_infos.as_ref().unwrap().len() as u32 {
+            return Err(Error::InvalidParam);
+        }
+        let pcm_info = &self.pcm_infos.as_ref().unwrap()[stream_id as usize];
+        Ok(pcm_info.features.into())
+    }
+
+    /// Get the latest notification.
+    pub fn latest_notification(&mut self) -> Result<Notification> {
+        if let Some(token) = self.event_queue.peek_used() {
+            unsafe {
+                self.event_queue
+                    .pop_used(token, &[], &mut [&mut self.event_buf])?;
+            }
+        } else {
+            info!("Not ready");
+            return Err(Error::NotReady);
+        }
+        let event = VirtIOSndEvent::read_from_prefix(&self.event_buf).unwrap();
+        Ok(Notification {
+            notification_type: NotificationType::from(event.hdr.command_code),
+            data: event.data,
+        })
+    }
+}
+
+/// The status of the PCM stream.
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
+enum PCMState {
+    #[default]
+    SetParams,
+    Prepare,
+    Release,
+    Start,
+    Stop,
+}
+
+/// Stream errors.
+#[derive(Debug, PartialEq, Eq)]
+enum StreamError {
+    InvalidState(&'static str, PCMState),
+    InvalidStateTransition(PCMState, PCMState),
+    InvalidStreamId(u32),
+    DescriptorReadFailed,
+    DescriptorWriteFailed,
+    CouldNotDisconnectStream,
+}
+
+macro_rules! set_new_state {
+    ($new_state_fn:ident, $new_state:expr, $($valid_source_states:tt)*) => {
+        fn $new_state_fn(&mut self) -> PCMStateResult<()> {
+            if !matches!(self, $($valid_source_states)*) {
+                return Err(StreamError::InvalidStateTransition(*self, $new_state));
+            }
+            *self = $new_state;
+            Ok(())
+        }
+    };
+}
+
+impl PCMState {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    set_new_state!(
+        set_parameters,
+        Self::SetParams,
+        Self::SetParams | Self::Prepare | Self::Release
+    );
+
+    set_new_state!(
+        prepare,
+        Self::Prepare,
+        Self::SetParams | Self::Prepare | Self::Release
+    );
+
+    set_new_state!(start, Self::Start, Self::Prepare | Self::Stop);
+
+    set_new_state!(stop, Self::Stop, Self::Start);
+
+    set_new_state!(release, Self::Release, Self::Prepare | Self::Stop);
```

Comment:
The methods generated by these macros don't seem to be called anywhere, what are they intended for?

Reply:
Now I've removed the unused macros.

Reply:
Do you actually use `pcm_states` anywhere?

Reply:
**Other places indeed do not use `pcm_states`**. These macros were originally intended to determine if the caller switched the PCM states in the correct order. However, due to time constraints, this state checking, which could have enhanced robustness, was not implemented.